### DISCUSSION TITLE
--strict to enforce static rules  in global scope

### DISCUSSION THREAD

**Discussion by gryznar** [modular/modular#411] [Ideas]
I am just wondering if it could be possible to have such option. Mojo as a superset of Python has to be compatible with its code, but it affects also global scope in which system programmers currently compromise due to Python dependency.

Adding handlig of such flag to compiler could enforce the same rules which happen in fn / structs, but in global scope and satisfy these 2 worlds. It should also translate into more aggressive optimizations and improve execution time. Will it be possible to do it in such way?

Personally also it will make me very glad, cause I love Python syntax, but I am not a big fan of its dynamism in which speed is sacrificed.

**Comment by gryznar** [comment]
This could be also compatible with Python world in which mypy has also strict mode. What do you think about that?

**Comment by Mogball** [comment]
Mojo's dynamism should be a "pay for what you use" kind of implementation. The existence of all the dynamic features does not affect the performance of generated code unless they are used. Note that the REPL environment is a bit different because it is the only Mojo environment that is inherently dynamic (as REPLs tend to be). A `--strict` flag would not boost performance, but would essentially just disable dynamic features. It's not clear what that means: would it just disable `def` syntax?

**Reply by gryznar** [reply]
Not exactly. It is just a way to enforce specifying type of variable and its mutability in global scope as in `fn` (static way). 

Currently it is not possible due to compatibility with Python in which all variables are dynamic, so there is not a way to point that global variable is not declared with `let` or `var` by the compiler.

**Reply by gryznar** [reply]
For a system programmers Mojo can be treated like a static language within classes (`struct` ) and within functions / methods (`fn`). It cannot be treated the same way only in global scope. This flag will allow to support static programming also in this area.

**Reply by ksandvik** [reply]
It's better to have possible specific directives that are mojo-specific for defining static functions et rest, i.e. fn that is static in nature. Anything else should be pythonic.

**Reply by Mogball** [reply]
@gryznar Ah sorry I misunderstood. Yes you are right that there is no way to control the semantics of allowed features in top-level code. It was indeed discussed to have this as a command-line flag, since there is no other way to control it.

**Reply by gryznar** [reply]
@Mogball That's right! I would like very much to see that Mojo will support static programming everywhere if a programmer will wish such. I love Python simple and intuitive syntax but static nature is very missing feature for me in Python... If I am not wrong it is not so hard to implement such option in compiler.

**Reply by gryznar** [reply]
As I have noted before, it is consistent with Python world in which static tools like mypy are very helpful. Mojo can go step further and introduce it to the compiler. 

I am proposing that, because static tools are not ideal solution here. There have to be updated, have bugs, not supports every feature etc. Implementing it as a part of language design could be much better

**Reply by gryznar** [reply]
@ksandvik linters are also very Pythonic :) Mojo could just do it via compiler, not via external tools, which may not be very up-to-date. Current dual nature (static and dynamic) does not work only in global scope. So it is kind of inconsistency  that there is no possibility to control that in top-level code

**Reply by Mogball** [reply]
> So it is kind of inconsistency that there is no possibility to control that in top-level code

Yeah this is a valid point. I believe the same topic was brought up somewhere else. The default right now is in a bit of a weird place. Mojo top-level code behaves differently than anything else. It is "strict" by default, however.

**Reply by gryznar** [reply]
So if topics which such requests are rising (which means that some people also want "static Python") and this is also consistent with the rest, maybe Mojo should introduce such flag?

**Reply by Mogball** [reply]
It should have one. The problem right now is that top-level Mojo code is neither "static" in the `fn` sense or "dynamic" in the `def` sense. The goal is to make it behave like an imperative body of code, in which case such a flag would have to be added.

**Reply by gryznar** [reply]
So maybe it would be nice to add that to the roadmap? For future clarity :)

**Reply by gryznar** [reply]
If that will be added, issue can be closed ;) @Mogball feel free to close it after that

**Reply by Mogball** [reply]
"What to do about top-level code" I think is worth a discussion before it gets thrown into the roadmap at this point :P

**Comment by ksandvik** [comment]
Top level fn functions?

**Reply by gryznar** [reply]
Currently there are supported, so I don't understand question
